### PR TITLE
feat: add getIdentityPermissionsByIdentity

### PR DESCRIPTION
**Pull request by htztomic**
Issue #, if available:

Description of changes:

Checklist:

* [ ] Have you successfully deployed to an AWS account with your changes?
* [ ] Have you written new tests for your core changes, as applicable?

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: workbench-core/authorization/README.md [hunk 1]
```diff
@@ -5,7 +5,11 @@
 # Code Coverage
 | Statements                  | Branches                | Functions                 | Lines             |
 | --------------------------- | ----------------------- | ------------------------- | ----------------- |
+<<<<<<< HEAD
+| ![Statements](https://img.shields.io/badge/statements-93.97%25-brightgreen.svg?style=flat) | ![Branches](https://img.shields.io/badge/branches-100%25-brightgreen.svg?style=flat) | ![Functions](https://img.shields.io/badge/functions-63.56%25-red.svg?style=flat) | ![Lines](https://img.shields.io/badge/lines-93.57%25-brightgreen.svg?style=flat) |
+=======
 | ![Statements](https://img.shields.io/badge/statements-93.24%25-brightgreen.svg?style=flat) | ![Branches](https://img.shields.io/badge/branches-100%25-brightgreen.svg?style=flat) | ![Functions](https://img.shields.io/badge/functions-61.78%25-red.svg?style=flat) | ![Lines](https://img.shields.io/badge/lines-92.77%25-brightgreen.svg?style=flat) |
+>>>>>>> 10c6b71473ef8e9e9c5b8387eccbe0d06b8f47c8
```
Comment (bhudnell) [inline workbench-core/authorization/README.md:12]: missed merge conflict

File: solutions/swb-reference/src/SWBStack.ts [hunk 2]
```diff
@@ -861,7 +861,7 @@ export class SWBStack extends Stack {
       gsis: [
         {
           indexName: 'getIdentityPermissionsByIdentity',
-          partitionKey: { name: 'Identity', type: AttributeType.STRING },
+          partitionKey: { name: 'identity', type: AttributeType.STRING },
```
Comment (bhudnell) [inline solutions/swb-reference/src/SWBStack.ts:864]: why this change?

File: workbench-core/example/app/src/routes/dynamicAuthorizationRoutes.ts [hunk 3]
```diff
@@ -170,4 +170,15 @@ export function setUpDynamicAuthorizationRoutes(router: Router, service: Dynamic
       }
     })
   );
+  router.get(
+    '/authorization/identitypermissions/identity',
```
Comment (bhudnell) [inline workbench-core/example/app/src/routes/dynamicAuthorizationRoutes.ts:174]: In #733 it was suggested to change to `/authorization/permissions/*`

File: workbench-core/authorization/src/dynamicAuthorization/dynamicAuthorizationService.test.ts [hunk 4]
```diff
@@ -357,4 +357,56 @@ describe('DynamicAuthorizationService', () => {
...
+    let sampleGroupType: IdentityType;
+
+    let sampleAction: Action;
+    let sampleEffect: Effect;
+    let sampleSubjectType: string;
+    let sampleSubjectId: string;
+    let sampleConditions: Record<string, JSONValue>;
+    let sampleFields: string[];
+    let sampleDescription: string;
+    let mockIdentityPermission: IdentityPermission;
```
Comment (bhudnell) [inline workbench-core/authorization/src/dynamicAuthorization/dynamicAuthorizationService.test.ts:372]: a lot of these look like duplicates of `createIdentityPermissions`. Can we extract them to a higher level block to reduce duplication?

File: workbench-core/authorization/src/dynamicAuthorization/dynamicAuthorizationInputs/getIdentityPermissionsByIdentity.ts [hunk 5]
```diff
@@ -3,27 +3,37 @@
  *  SPDX-License-Identifier: Apache-2.0
  */
 
-import { IdentityPermission, IdentityType } from './identityPermission';
+import { z } from 'zod';
+import { IdentityPermission, IdentityTypeParser } from './identityPermission';
 
+// eslint-disable-next-line @rushstack/typedef-var
+export const GetIdentityPermissionsByIdentityRequestParser = z
```
Comment (bhudnell) [inline workbench-core/authorization/src/dynamicAuthorization/dynamicAuthorizationInputs/getIdentityPermissionsByIdentity.ts:10]: Do we need request parsers at the service level?

File: workbench-core/authorization/src/dynamicAuthorization/ddbDynamicAuthorizationPermissionsPlugin.ts [hunk 6]
```diff
@@ -163,4 +188,49 @@ export class DDBDynamicAuthorizationPermissionsPlugin implements DynamicAuthoriz
   ): string {
     return `${action}|${effect}|${identityType}|${identityId}`;
   }
+  private _transformItemToIdentityPermission(item: Record<string, JSONValue>): IdentityPermission {
+    const { pk, sk, conditions, fields, description } = item;
```
Comment (bhudnell) [inline workbench-core/authorization/src/dynamicAuthorization/ddbDynamicAuthorizationPermissionsPlugin.ts:192]: What happens if these fields dont exist in the `item` object?

File: workbench-core/example/app/src/routes/dynamicAuthorizationRoutes.ts (hunk 3 above)
Comment (DaveMadsen) [reply workbench-core/example/app/src/routes/dynamicAuthorizationRoutes.ts:174]: It was a suggestion. Not sure if there is any reason why permissions would get further overloaded by some other entity.

File: solutions/swb-reference/src/SWBStack.ts (hunk 2 above)
Comment (insignias) [reply solutions/swb-reference/src/SWBStack.ts:864]: This is a part of this [PR](https://github.com/aws-solutions/solution-spark-on-aws/pull/732/files#diff-d52cbf56a77bfc71120b1669d3a69b00dcc221ef47f74d4b7b34cdd0f6d9fc41) and needs to deployed to our test account first as mentioned [here](https://github.com/aws-solutions/solution-spark-on-aws/pull/732#pullrequestreview-1234866718)

Comment (bhudnell) [reply solutions/swb-reference/src/SWBStack.ts:864]: The comment there makes it seem like the change didnt originate there?

Comment (insignias) [reply solutions/swb-reference/src/SWBStack.ts:864]: The DynamicAuth DynamoDB GSI was created with partitionKey `Identity`, @htztomic pointed out it should have been `identity`. In order to deploy this change we first need to destroy the DDBTable and then recreate it with correct GSI parameters. Hence it would be better to handle this change in a single PR where we update both the SWBStack and the ExampleStack together.

Comment (bhudnell) [reply solutions/swb-reference/src/SWBStack.ts:864]: gotcha, makes sense

File: workbench-core/authorization/src/dynamicAuthorization/ddbDynamicAuthorizationPermissionsPlugin.ts (hunk 6 above)
Comment (htztomic) [reply workbench-core/authorization/src/dynamicAuthorization/ddbDynamicAuthorizationPermissionsPlugin.ts:192]: these items are transformed from identity permissions. These values are included in the transactWrite when a permission is written

File: workbench-core/authorization/src/dynamicAuthorization/dynamicAuthorizationInputs/getIdentityPermissionsByIdentity.ts (hunk 5 above)
Comment (htztomic) [reply workbench-core/authorization/src/dynamicAuthorization/dynamicAuthorizationInputs/getIdentityPermissionsByIdentity.ts:10]: Yes, it is being inferred into the type. This can be utilized to validate input if necessary.

File: workbench-core/authorization/src/dynamicAuthorization/ddbDynamicAuthorizationPermissionsPlugin.ts [hunk 7]
```diff
@@ -60,7 +62,30 @@ export class DDBDynamicAuthorizationPermissionsPlugin implements DynamicAuthoriz
...
+    const queryParams = {
+      index: this._getIdentityPermissionsByIdentityIndex,
+      key: {
+        name: this._getIdentityPermissionsByIdentityPartitionKey,
+        value: identity
+      }
+    };
+    let { data, paginationToken } = await this._dynamoDBService.getPaginatedItems(queryParams);
+
+    while (paginationToken) {
```
Comment (aws-tyler) [inline workbench-core/authorization/src/dynamicAuthorization/ddbDynamicAuthorizationPermissionsPlugin.ts:76]: This isn't something I'm super concerned about but want to get on your radar: This is a potential DoS issue, but only at scale. If someone accidentally created enough permissions, they could time out the call by making this loop run for longer than the response timeout. There is possibly also a response-body-size related strangling, but with serverless architecture that may be less of an issue. Anyway, not something I'm worried about for now, but if there is a way to return this data in a paginated way, we could avoid the issue completely.

File: workbench-core/authorization/src/dynamicAuthorization/dynamicAuthorizationService.ts [hunk 8]
```diff
@@ -229,6 +234,23 @@ export class DynamicAuthorizationService {
...
+  /**
+   * Get all identity permissions associated to a specific identity
+   * @param getIdentityPermissionsByIdentityRequest - {@link GetIdentityPermissionsByIdentityRequest}
+   *
+   * @returns - {@link GetIdentityPermissionsByIdentityResponse}
+   */
+  public async getIdentityPermissionsByIdentity(
+    getIdentityPermissionsByIdentityRequest: GetIdentityPermissionsByIdentityRequest
+  ): Promise<GetIdentityPermissionsByIdentityResponse> {
+    getIdentityPermissionsByIdentityRequest = GetIdentityPermissionsByIdentityRequestParser.parse(
```
Comment (aws-tyler) [inline workbench-core/authorization/src/dynamicAuthorization/dynamicAuthorizationService.ts:246]: To help delineate when an item has been parsed/validated and when it hasn't, we've been using `validatedRequest`. Not a blocker though.

File: solutions/swb-reference/src/SWBStack.ts (hunk 2 above)
Comment (aws-tyler) [reply solutions/swb-reference/src/SWBStack.ts:864]: Is this going to be a breaking change for anything that has been deployed recently?

**Review by aws-tyler — DISMISSED** [review]

File: workbench-core/authorization/src/dynamicAuthorization/ddbDynamicAuthorizationPermissionsPlugin.ts (hunk 6 above)
Comment (bhudnell) [reply workbench-core/authorization/src/dynamicAuthorization/ddbDynamicAuthorizationPermissionsPlugin.ts:192]: Since it looks like these values are required, what are your thoughts on turning this into a more explicit interface?

Comment (aws-tyler) [reply workbench-core/authorization/src/dynamicAuthorization/ddbDynamicAuthorizationPermissionsPlugin.ts:192]: Zod could also do parsing/validation check to ensure everything is present and the expected type, which would help get rid of those `as` statements

File: workbench-core/authorization/src/dynamicAuthorization/dynamicAuthorizationService.test.ts [hunk 9]
```diff
@@ -454,6 +458,20 @@ describe('DynamicAuthorizationService', () => {
...
+      }
+    });
+    const request = {
+      identityId: sampleGroupId,
+      identityType: sampleGroupType
+    };
+
+    const { data } = await dynamicAuthzService.getIdentityPermissionsByIdentity(request);
+    expect(data.identityPermissions).toStrictEqual([mockIdentityPermission]);
+  });
```
Comment (bhudnell) [inline workbench-core/authorization/src/dynamicAuthorization/dynamicAuthorizationService.test.ts:474]: I think this test got separated from its parent `describe` in the merge from develop

File: workbench-core/authorization/src/dynamicAuthorization/dynamicAuthorizationService.test.ts [hunk 10]
```diff
@@ -202,45 +239,12 @@ describe('DynamicAuthorizationService', () => {
...
-    let sampleFields: string[];
-    let sampleDescription: string;
-
-    let mockIdentityPermission: IdentityPermission;
-
-    let mockIdentityPermissions: IdentityPermission[];
+    let auditServiceWriteSpy: jest.SpyInstance;
+    beforeAll(() => {
+      auditServiceWriteSpy = jest.spyOn(auditService, 'write');
+    });
```
Comment (bhudnell) [inline workbench-core/authorization/src/dynamicAuthorization/dynamicAuthorizationService.test.ts:245]: This spy has been moved to the outermost `describe` block

File: workbench-core/authorization/src/dynamicAuthorization/dynamicAuthorizationService.ts (hunk 8 above)
Comment (bhudnell) [inline workbench-core/authorization/src/dynamicAuthorization/dynamicAuthorizationService.ts:246]: Im not sure its necessary to use zod to validate function params at the service level, but not a blocker

Comment (aws-tyler) [reply workbench-core/authorization/src/dynamicAuthorization/dynamicAuthorizationService.ts:246]: This is probably way too deep of a comment but 🤷 haha. @htztomic and i were actually discussing this the other day. I think TypeScript's lack of runtime validation is a gaping hole in their "we're a typed language!" pitch and so its worth thinking about validation a little differently than you would with Java or something (where you only need logical validation and not typing validation). Validation should probably happen at all module boundaries because someone could be REALLY COOL and use an `as` or something to override the type system's static check and then break things at run time. Zod helps us close that gap a little bit by doing the runtime check via parsing.

If this were just a controller and service that were paired together and never being reused, then ya its probably overkill, but i think validation at the level the module with interface with other code is good to help us make sure our data is always what we expect it to be at run time. Thoughts?

File: workbench-core/authorization/src/dynamicAuthorization/ddbDynamicAuthorizationPermissionsPlugin.ts [hunk 11]
```diff
@@ -163,4 +188,49 @@ export class DDBDynamicAuthorizationPermissionsPlugin implements DynamicAuthoriz
...
+      fields: fields as string[],
+      description: description as string
+    };
+  }
+
+  private _decomposeIdentityPermissionsPartitionKey(partitionKey: string): {
+    subjectType: string;
+    subjectId: string;
+  } {
+    const values = partitionKey.split('|');
```
[the remark this answers is not in the document]
Comment (htztomic) [reply workbench-core/authorization/src/dynamicAuthorization/ddbDynamicAuthorizationPermissionsPlugin.ts:215]: delimiter shouldn't change but I can abstract it

File: workbench-core/authorization/src/dynamicAuthorization/ddbDynamicAuthorizationPermissionsPlugin.ts (hunk 7 above)
Comment (htztomic) [reply workbench-core/authorization/src/dynamicAuthorization/ddbDynamicAuthorizationPermissionsPlugin.ts:76]: Yeah the DoS attack is interesting, these APIs should not be exposed at the handler level directly anyway.

Comment (aws-tyler) [reply workbench-core/authorization/src/dynamicAuthorization/ddbDynamicAuthorizationPermissionsPlugin.ts:76]: API exposure isn't the problem, its the amount of data being retrieved. We're ok for small sets of data, but we'd e ok for big sets of data too if we could return just the page of results instead of the entire aggregation of all results.

Comment (htztomic) [reply workbench-core/authorization/src/dynamicAuthorization/ddbDynamicAuthorizationPermissionsPlugin.ts]: added paginated results

**Review by aws-tyler — APPROVED** [review]

File: workbench-core/authorization/src/dynamicAuthorization/ddbDynamicAuthorizationPermissionsPlugin.test.ts [hunk 12]
```diff
@@ -244,11 +234,20 @@ describe('DDB Dynamic Authorization Permissions Plugin tests', () => {
...
         });
 
-      const { data } = await dynamoDBDynamicPermissionsPlugin.getIdentityPermissionsByIdentity({
+      const { data, paginationToken } =
+        await dynamoDBDynamicPermissionsPlugin.getIdentityPermissionsByIdentity({
+          identityId: sampleGroupId,
+          identityType: sampleGroupType
+        });
+      expect(data.identityPermissions).toStrictEqual([mockIdentityPermission]);
+      expect(paginationToken).toStrictEqual(base64PaginationToken);
```
Comment (bhudnell) [inline workbench-core/authorization/src/dynamicAuthorization/ddbDynamicAuthorizationPermissionsPlugin.test.ts:243]: I'd recommend testing that the entire received response matches the expected response object rather than testing each field separately. This will help ensure we update tests if we add more fields to the response object

Comment (htztomic) [reply workbench-core/authorization/src/dynamicAuthorization/ddbDynamicAuthorizationPermissionsPlugin.test.ts:243]: Each field can be different per response while others can stay the same, testing an entire response can be a plus but I don't think it is necessary here

**Review by bhudnell — APPROVED** [review]
